# Data Merging and Organization

The purpose of this notebook is to upload and organize data from the table generator tool and split it up by years. Each set of years is stored in a different dataframe and exported to a .csv file.

### Uploading Data
Data had to be obtained in 7 separate parts because the table generator tool only allowed for a max of 75 columns per session. 
The goal now is to merge all files together by 'School ID - NCES Assigned [Public School] Latest available year' and delete repeated columns: 'School Name', and 'State Name [Public School] Latest available year'

In [48]:
import pandas as pd

# Add data files to team project directory and check
#%ls

# Upload the 7 data files
part1 = pd.read_csv('all_years_1.csv', na_values = '-')
part2 = pd.read_csv('all_years_2.csv', na_values = '-')
part3 = pd.read_csv('all_years_3.csv', na_values = '-')
part4 = pd.read_csv('all_years_4.csv', na_values = '-')
part5 = pd.read_csv('all_years_5.csv', na_values = '-')
part6 = pd.read_csv('all_years_6.csv', na_values = '-')
part7 = pd.read_csv('all_years_7.csv', na_values = '-')

# Drop 'School Name', and 'State Name [Public School] Latest available year' columns
part2.drop(columns = ['School Name','State Name [Public School] Latest available year'])
part3.drop(columns = ['School Name','State Name [Public School] Latest available year'])
part4.drop(columns = ['School Name','State Name [Public School] Latest available year'])
part5.drop(columns = ['School Name','State Name [Public School] Latest available year'])
part6.drop(columns = ['School Name','State Name [Public School] Latest available year'])
part7.drop(columns = ['School Name','State Name [Public School] Latest available year'])

# Merge by 'School ID - NCES Assigned [Public School] Latest available year'
merged = part1.merge(part2, how='outer', on='School ID - NCES Assigned [Public School] Latest available year')
merged = merged.merge(part3, how='outer', on='School ID - NCES Assigned [Public School] Latest available year')
merged = merged.merge(part4, how='outer', on='School ID - NCES Assigned [Public School] Latest available year')
merged = merged.merge(part5, how='outer', on='School ID - NCES Assigned [Public School] Latest available year')
merged = merged.merge(part6, how='outer', on='School ID - NCES Assigned [Public School] Latest available year')
merged = merged.merge(part7, how='outer', on='School ID - NCES Assigned [Public School] Latest available year')
merged.info()
merged.head()

C:\Users\yessi\anaconda3\lib\site-packages\pandas\core\frame.py:9190: FutureWarning: Passing 'suffixes' which cause duplicate columns {'School Name_x', 'State Name [Public School] Latest available year_x'} in the result is deprecated and will raise a MergeError in a future version.
  return merge(


<class 'pandas.core.frame.DataFrame'>
Int64Index: 68817 entries, 0 to 68816
Columns: 424 entries, School Name_x to Free Lunch Eligible [Public School] 1987-88
dtypes: object(424)
memory usage: 223.1+ MB


School Name_x State Name [Public School] Latest available year_x  \
0      ="03-MM"                                            Florida   
1     113-H-W-4                                            FLORIDA   
2    118-E-SW-5                                            FLORIDA   
3      12-E-C-1                                            Florida   
4     132-M-W-4                                            FLORIDA   

  School ID - NCES Assigned [Public School] Latest available year  \
0                                       120150006730                
1                                       120144008827                
2                                       120144008672                
3                                       120144007627                
4                                       120144008872                

  Agency ID - NCES Assigned [Public School] Latest available year  \
0                                            1201500                
1                                            1201440                
2                                            1201440                
3                                            1201440                
4                                            1201440                

  County Number [Public School] 2019-20 County Name [Public School] 2019-20  \
0                                     †                                   †   
1                                 12095                       Orange County   
2                                     †                                   †   
3                                     †                                   †   
4                                 12095                       Orange County   

  School Type [Public School] 2019-20 Charter School [Public School] 2019-20  \
0                                   †                                      †   
1                    1-Regular school                                   2-No   
2                                   †                                      †   
3                                   †                                      †   
4                    1-Regular school                                   2-No   

  Magnet School [Public School] 2019-20  \
0                                     †   
1                                     †   
2                                     †   
3                                     †   
4                                     †   

  Shared Time School [Public School] 2019-20  ...  \
0                                          †  ...   
1                                          †  ...   
2                                          †  ...   
3                                          †  ...   
4                                          †  ...   

  Free Lunch Eligible [Public School] 1996-97  \
0                                           †   
1                                           †   
2                                           †   
3                                           †   
4                                           †   

  Free Lunch Eligible [Public School] 1995-96  \
0                                           †   
1                                           †   
2                                           †   
3                                           †   
4                                           †   

  Free Lunch Eligible [Public School] 1994-95  \
0                                           †   
1                                           †   
2                                           †   
3                                           †   
4                                           †   

  Free Lunch Eligible [Public School] 1993-94  \
0                                           †   
1                                           †   
2                                           †   
3                                           †   
4                                           †   

  Free Lunch Eligible [Public School]

In [47]:
# create new csv file with entire dataset
merged.to_csv('all_data.csv')

# list of all column names for each year in data set
my_list = list(merged)
print (my_list)

['School Name_x', 'State Name [Public School] Latest available year_x', 'School ID - NCES Assigned [Public School] Latest available year', 'Agency ID - NCES Assigned [Public School] Latest available year', 'County Number [Public School] 2019-20', 'County Name [Public School] 2019-20', 'School Type [Public School] 2019-20', 'Charter School [Public School] 2019-20', 'Magnet School [Public School] 2019-20', 'Shared Time School [Public School] 2019-20', 'Locale [Public School] 2005-06', 'School-wide Title I [Public School] 2019-20', 'School-wide Title I [Public School] 2018-19', 'School-wide Title I [Public School] 2017-18', 'School-wide Title I [Public School] 2016-17', 'School-wide Title I [Public School] 2015-16', 'School-wide Title I [Public School] 2014-15', 'School-wide Title I [Public School] 2013-14', 'School-wide Title I [Public School] 2012-13', 'School-wide Title I [Public School] 2011-12', 'School-wide Title I [Public School] 2010-11', 'School-wide Title I [Public School] 2009-

### Separate Data by Year

We can see above that there are  424 different columns. To organize the data we will split the data into 5 sections: Ground Truth Labeling, Feature Selection, Modeling, Model Selection, and Test data. Since Title 1 designation is not available for the years prior to the 1998-1999 school year, we will use school years 1998-2003 for Ground Truth Labeling, 2003-2008 for Feature Selection, and 2008-2018 for Modeling for the Midterm Check In and will make adjustments for adjustments for the project submission.

In [49]:
# create new dataframe for years 1998-2003
all_years = ['School Name', 'State Name [Public School] Latest available year', 
             'School ID - NCES Assigned [Public School] Latest available year', 
             'Agency ID - NCES Assigned [Public School] Latest available year', 'County Number [Public School] 2019-20']
year1998 = [col for col in merged.columns if '1998-99' in col]
year1999 = [col for col in merged.columns if '1999-00' in col]
year2000 = [col for col in merged.columns if '2000-01' in col]
year2001 = [col for col in merged.columns if '2001-02' in col]
year2002 = [col for col in merged.columns if '2002-03' in col]

total = all_years + year1998 + year1999 + year2000 + year2001 + year2002
years_1998_2003 = merged[total]
groundtruth = years_1998_2003.copy()
groundtruth.info()
groundtruth.to_csv('groundtruth_98_to_03.csv')
groundtruth.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68817 entries, 0 to 68816
Data columns (total 70 columns):
 #   Column                                                            Non-Null Count  Dtype 
---  ------                                                            --------------  ----- 
 0   School Name                                                       68013 non-null  object
 1   State Name [Public School] Latest available year                  51629 non-null  object
 2   School ID - NCES Assigned [Public School] Latest available year   52433 non-null  object
 3   Agency ID - NCES Assigned [Public School] Latest available year   52433 non-null  object
 4   County Number [Public School] 2019-20                             52433 non-null  object
 5   School-wide Title I [Public School] 1998-99                       52433 non-null  object
 6   Total Students All Grades (Excludes AE) [Public School] 1998-99   52433 non-null  object
 7   Free and Reduced Lunch Students [Public 

School Name State Name [Public School] Latest available year  \
0    ="03-MM"                                          Florida   
1   113-H-W-4                                          FLORIDA   
2  118-E-SW-5                                          FLORIDA   
3    12-E-C-1                                          Florida   
4   132-M-W-4                                          FLORIDA   

  School ID - NCES Assigned [Public School] Latest available year  \
0                                       120150006730                
1                                       120144008827                
2                                       120144008672                
3                                       120144007627                
4                                       120144008872                

  Agency ID - NCES Assigned [Public School] Latest available year  \
0                                            1201500                
1                                            1201440                
2                                            1201440                
3                                            1201440                
4                                            1201440                

  County Number [Public School] 2019-20  \
0                                     †   
1                                 12095   
2                                     †   
3                                     †   
4                                 12095   

  School-wide Title I [Public School] 1998-99  \
0                                           †   
1                                           †   
2                                           †   
3                                           †   
4                                           †   

  Total Students All Grades (Excludes AE) [Public School] 1998-99  \
0                                                  †                
1                                                  †                
2                                                  †                
3                                                  †                
4                                                  †                

  Free and Reduced Lunch Students [Public School] 1998-99  \
0                                                  †        
1                                                  †        
2                                                  †        
3                                                  †        
4                                                  †        

  Male Students [Public School] 1998-99  \
0                                     †   
1                                     †   
2                                     †   
3                                     †   
4                                     †   

  Female Students [Public School] 1998-99  ...  \
0                                       †  ...   
1                                       †  ...   
2                                       †  ...   
3                                       †  ...   
4                                       †  ...   

  Male Students [Public School] 2002-03  \
0                                     †   
1                                     †   
2                                     †   
3                                     †   
4                                     †   

  Female Students [Public School] 2002-03  \
0                                       †   
1                                       †   
2                                       †   
3                                       †   
4                                       †   

  American Indian/Alaska Native Students [Public School] 2002-03  \
0                                                  †               
1                                                  †               
2                                                  †               
3                                                  †               
4        

In [50]:
# create new dataframe for years 2003-2008
year2003 = [col for col in merged.columns if '2003-04' in col]
year2004 = [col for col in merged.columns if '2004-05' in col]
year2005 = [col for col in merged.columns if '2005-06' in col]
year2006 = [col for col in merged.columns if '2006-07' in col]
year2007 = [col for col in merged.columns if '2007-08' in col]

total = all_years + year2003 + year2004 + year2005 + year2006 + year2007
years_2003_2008 = merged[total]
feature_selection = years_2003_2008.copy()
feature_selection.info()
groundtruth.to_csv('feature_selection_03_to_08.csv')
groundtruth.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68817 entries, 0 to 68816
Data columns (total 71 columns):
 #   Column                                                            Non-Null Count  Dtype 
---  ------                                                            --------------  ----- 
 0   School Name                                                       68013 non-null  object
 1   State Name [Public School] Latest available year                  51629 non-null  object
 2   School ID - NCES Assigned [Public School] Latest available year   52433 non-null  object
 3   Agency ID - NCES Assigned [Public School] Latest available year   52433 non-null  object
 4   County Number [Public School] 2019-20                             52433 non-null  object
 5   School-wide Title I [Public School] 2003-04                       52433 non-null  object
 6   Total Students All Grades (Excludes AE) [Public School] 2003-04   52433 non-null  object
 7   Free and Reduced Lunch Students [Public 

School Name State Name [Public School] Latest available year  \
0    ="03-MM"                                          Florida   
1   113-H-W-4                                          FLORIDA   
2  118-E-SW-5                                          FLORIDA   
3    12-E-C-1                                          Florida   
4   132-M-W-4                                          FLORIDA   

  School ID - NCES Assigned [Public School] Latest available year  \
0                                       120150006730                
1                                       120144008827                
2                                       120144008672                
3                                       120144007627                
4                                       120144008872                

  Agency ID - NCES Assigned [Public School] Latest available year  \
0                                            1201500                
1                                            1201440                
2                                            1201440                
3                                            1201440                
4                                            1201440                

  County Number [Public School] 2019-20  \
0                                     †   
1                                 12095   
2                                     †   
3                                     †   
4                                 12095   

  School-wide Title I [Public School] 1998-99  \
0                                           †   
1                                           †   
2                                           †   
3                                           †   
4                                           †   

  Total Students All Grades (Excludes AE) [Public School] 1998-99  \
0                                                  †                
1                                                  †                
2                                                  †                
3                                                  †                
4                                                  †                

  Free and Reduced Lunch Students [Public School] 1998-99  \
0                                                  †        
1                                                  †        
2                                                  †        
3                                                  †        
4                                                  †        

  Male Students [Public School] 1998-99  \
0                                     †   
1                                     †   
2                                     †   
3                                     †   
4                                     †   

  Female Students [Public School] 1998-99  ...  \
0                                       †  ...   
1                                       †  ...   
2                                       †  ...   
3                                       †  ...   
4                                       †  ...   

  Male Students [Public School] 2002-03  \
0                                     †   
1                                     †   
2                                     †   
3                                     †   
4                                     †   

  Female Students [Public School] 2002-03  \
0                                       †   
1                                       †   
2                                       †   
3                                       †   
4                                       †   

  American Indian/Alaska Native Students [Public School] 2002-03  \
0                                                  †               
1                                                  †               
2                                                  †               
3                                                  †               
4        

In [51]:
# create new dataframe for years 2008-2018
year2008 = [col for col in merged.columns if '2008-09' in col]
year2009 = [col for col in merged.columns if '2009-10' in col]
year2010 = [col for col in merged.columns if '2010-11' in col]
year2011 = [col for col in merged.columns if '2011-12' in col]
year2012 = [col for col in merged.columns if '2012-13' in col]
year2013 = [col for col in merged.columns if '2013-14' in col]
year2014 = [col for col in merged.columns if '2014-15' in col]
year2015 = [col for col in merged.columns if '2015-16' in col]
year2016 = [col for col in merged.columns if '2016-17' in col]
year2017 = [col for col in merged.columns if '2017-18' in col]
year2018 = [col for col in merged.columns if '2018-19' in col]

total = (all_years + year2008 + year2009 + year2010 + year2011 + year2012 + year2013 + year2014 + year2015 + year2016 +
        year2017 + year2018)
years_2008_2018 = merged[total]
modeling = years_2008_2018.copy()
modeling.info()
modeling.to_csv('modeling_2008_to_2018.csv')
modeling.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68817 entries, 0 to 68816
Columns: 170 entries, School Name to Free Lunch Eligible [Public School] 2018-19
dtypes: object(170)
memory usage: 89.8+ MB


School Name State Name [Public School] Latest available year  \
0    ="03-MM"                                          Florida   
1   113-H-W-4                                          FLORIDA   
2  118-E-SW-5                                          FLORIDA   
3    12-E-C-1                                          Florida   
4   132-M-W-4                                          FLORIDA   

  School ID - NCES Assigned [Public School] Latest available year  \
0                                       120150006730                
1                                       120144008827                
2                                       120144008672                
3                                       120144007627                
4                                       120144008872                

  Agency ID - NCES Assigned [Public School] Latest available year  \
0                                            1201500                
1                                            1201440                
2                                            1201440                
3                                            1201440                
4                                            1201440                

  County Number [Public School] 2019-20  \
0                                     †   
1                                 12095   
2                                     †   
3                                     †   
4                                 12095   

  School-wide Title I [Public School] 2008-09  \
0                                           †   
1                                           †   
2                                           †   
3                                           †   
4                                           †   

  Total Students All Grades (Excludes AE) [Public School] 2008-09  \
0                                                  †                
1                                                  †                
2                                                  †                
3                                                  †                
4                                                  †                

  Free and Reduced Lunch Students [Public School] 2008-09  \
0                                                  †        
1                                                  †        
2                                                  †        
3                                                  †        
4                                                  †        

  Male Students [Public School] 2008-09  \
0                                     †   
1                                     †   
2                                     †   
3                                     †   
4                                     †   

  Female Students [Public School] 2008-09  ...  \
0                                       †  ...   
1                                       †  ...   
2                                       †  ...   
3                                       †  ...   
4                                       †  ...   

  American Indian/Alaska Native Students [Public School] 2018-19  \
0                                                  †               
1                                                  †               
2                                                  †               
3                                                  †               
4                                                  †               

  Asian or Asian/Pacific Islander Students [Public School] 2018-19  \
0                                                  †                 
1                                                  †                 
2                                                  †                 
3                                                  †                 
4                                                  †                 

  Hispanic Students [Public School] 2018-19  \
0 

In [40]:
# create new dataframe for years 2014-2019
#year2014 = [col for col in merged.columns if '2014-15' in col]
#year2015 = [col for col in merged.columns if '2015-16' in col]
#year2016 = [col for col in merged.columns if '2016-17' in col]
#year2017 = [col for col in merged.columns if '2017-18' in col]
#year2018 = [col for col in merged.columns if '2018-19' in col]

#total = all_years + year2014 + year2015 + year2016 + year2017 + year2018
#years_2014_2019 = merged[total]
#model_selection = years_2014_2019.copy()
#model_selection.info()
#model_selection.head(10)
#model_selection.to_csv('model_selection_start_to_end.csv')

In [39]:
# create new dataframe for years 2019-2021 (Note 2020-2021 data is not available)
#year2019 = [col for col in merged.columns if '2019-20' in col]

#total = all_years + year2019
#year_2019 = merged[total]
#test = year_2019.copy()
#test.info()
#test.head(10)
#test.to_csv('test_data_start_to_end.csv')